scarlet2/scene.py

```diff
@@ -175,23 +175,32 @@ def update_fn(updates, state, params):
         with tqdm.trange(max_iter, disable=not progress_bar) as t:
             for step in t:
                 # optimizer step
-                scene_, loss, opt_state = _make_step(scene, observations, optim, opt_state, filter_spec=filter_spec,
+                scene_, loss, opt_state, updates = _make_step(scene, observations, optim, opt_state, filter_spec=filter_spec,
                                                      constraint_fn=constraint_fn)
                 # Log the loss in the tqdm progress bar
                 t.set_postfix(loss=f"{loss:08.2f}")
 
                 # report current iteration results to callback
                 if callback is not None:
                     callback(scene_, loss)
-
-                # terminate optimization if all parameter change less than e_rel
+                
+                # terminate optimization if all gradient change less than e_rel
                 if e_rel is not None:
-                    crit = lambda x, x_: jnp.linalg.norm(x - x_) < e_rel * jnp.linalg.norm(x_)
-                    converged = tuple(
-                        crit(p, p_) for (p, p_) in zip(scene.parameters.values(), scene_.parameters.values()))
-                    if all(converged):
+                    crit = lambda x, x_: jnp.linalg.norm(x) < e_rel * jnp.linalg.norm(x_)
```

Comment:
You are assuming a specific set of `Parameter` instances to be given (spectrum and morphology in this case), but this may not be the case (or some of them may be frozen). I suggest to do this crit check more generally, namely by using `jax.tree_util.tree_reduce`, so that you can go through all of the pytree leaves automatically.